### DISCUSSION TITLE
Add possibility to send errors and exceptions to email

### DISCUSSION THREAD

**Discussion by PavelVsl** [oqtane/oqtane.framework#2014] [Ideas] [answered]
This is important for developer maintaining custom solution based on Oqtane. 

Solution: 
1.  On site or server level set account, group or host user for notifications. 
2. Log exceptions with full stack info 
3. Create notification with exception info and other debug information
4. Let system send this notification to developer

**Comment by sbwalker** [comment, marked as answer]
Oqtane already has a Logging service which captures exception information and other request context info. It also already has a Notification service which allows the system to send asynchronous emails to users or roles. So the only thing which is missing is a configuration option to specify that notifications should be sent for all log events of a certain Level or higher (ie. Error). And it would make sense for these notifications to be sent to anyone in the Host Users role, as these are the only users in the system which have the ability to view/manage system level information. I am also not sure if it would be a good idea to include the exception details in the notification itself - it may be a better idea to include a link where the host user can login to the site securely and view the exception details (to reduce disclosure risks).

**Reply by sbwalker** [reply]
#2022 includes the functionality described above. Based on the fact that logging is configured at the system level, a new notification option has been included at the system level as well. The option is surfaced through the System Info area of the Admin Dashboard and allows a host user to specify the minimum log level for which to send notifications (the default is Error and the feature is enabled by default). When an event is logged which is equal to or greater than the notification level specified, a notification is sent to all host users. The notification includes a subject indicating the level of the log event which occurred and provides a direct Url so the user can visit the Event Log and view more details. 

This feature is very valuable in terms of making the owner of an Oqtane installation aware of any serious errors which may be occurring in a run-time environment. However it should also be noted that this feature has the potential to be abused. For example, if a bot was able to trigger an error in a repetitive manner, the host users Inbox would get flooded with messages. Obviously the host user can mitigate this by setting the notification level to None to prevent further notifications from being sent while they investigate the issue further.

**Reply by markdav-is** [reply]
I'm investigating a bug where we continue to get errors emails to us even when the site logging level is set to none.  We keep checking the setting the option, but either the setting is not sticking or being ignored.

**Reply by sbwalker** [reply]
@markdav-is The minimum log level setting is currently stored in the appsettings.json file - so it affects all sites within an installation. However, this means that if you change the value of the setting you may need to restart the application for it to take effect. Have you verified that the value is set correctly in your appsettings.json in the environment? And has the app been restarted since the change?

**Reply by markdav-is** [reply]
Thanks for the note.  I was setting it using the UI under System Info - Options - Notifications = None (see attached screenshot) and I just verified that the server settings match (json below).  So I'm going to look at the notification code is see if there is some logic setting that's off.  I'll also restart the app service.

```
"LogLevel": {
      "Default": "Information",
      "Notify": "None"
    }

```
![image](https://user-images.githubusercontent.com/311063/175788745-9d478175-9105-48d2-850d-4dd38483a3cf.png)

**Reply by markdav-is** [reply]
We are seeing this bug most days.  I'm going to start an issue as to not spam up this thread.  I'm motivated to find it and fix it.  https://github.com/oqtane/oqtane.framework/issues/2259

**Comment by curlyfro** [comment]
this would be a great feature to have!

@chlupac are you suggesting emails should get sent to the module developer of custom modules?  if so, Oqtane would need to store the email addresses of the module maintainer which I don't think is there now and would need to be added to the db.  personally, I think this is something that should be managed by the module.  the module can tap into the logging and notification services, manage its own email addresses, and it can provide richer error messages.

it'd be nice if core modules could opt-in to this feature via Settings or something.

**Reply by PavelVsl** [reply]
I'm also thinking about this.  But sending messages to anyone out of the system should be permitted/confirmed by administrator. I like how this is done in Rider, you can preview which info is send to developers and you have to confirm it.
